### PR TITLE
Updated RecyclerView Adapter to List Adapter

### PR DESCRIPTION
**Pull request by tritran206**
- Updated RecyclerView Adapter to ListAdapters for LiveData and DiffUtil
- Change ViewModel factory to return same instance of ViewModel
- Reformat Code (created binding functions and moved functions into OnViewCreated in all fragments)
- Added Swipe Refresh to ProductDetailFragment
- Replaced TextInputLayout with TextInputEditText

### CODE REVIEW

File: app/src/main/java/com/example/wishlist/adapters/ProductAdapter.kt [hunk 1]
```diff
@@ -33,11 +35,15 @@ class ProductAdapter(val productList: List<Product>, val listener: OnProductClic
...
-
+private class ProductDiffCallback: DiffUtil.ItemCallback<Product>(){
+    override fun areItemsTheSame(oldItem: Product, newItem: Product): Boolean {
+        return oldItem.id == newItem.id
     }
 
-    override fun getItemCount(): Int {
-        return productList.size
+    override fun areContentsTheSame(oldItem: Product, newItem: Product): Boolean {
+        return oldItem == newItem
```
Comment (Tonyyytran) [inline app/src/main/java/com/example/wishlist/adapters/ProductAdapter.kt:48]: Does this check all attributes of the product?

File: app/src/main/java/com/example/wishlist/adapters/ProductAdapter.kt [hunk 2]
```diff
@@ -2,13 +2,15 @@ package com.example.wishlist
```
Comment (Tonyyytran) [inline app/src/main/java/com/example/wishlist/adapters/ProductAdapter.kt:2]: Can you update the package name?

File: app/src/main/java/com/example/wishlist/adapters/ReviewAdapter.kt [hunk 3]
```diff
@@ -2,12 +2,14 @@ package com.example.wishlist.data.model
```
Comment (Tonyyytran) [inline app/src/main/java/com/example/wishlist/adapters/ReviewAdapter.kt:2]: Same as above

File: app/src/main/java/com/example/wishlist/fragments/AddReviewFragment.kt [hunk 4]
```diff
@@ -8,9 +8,13 @@ import android.view.ViewGroup
 import android.widget.Toast
 import androidx.lifecycle.ViewModelProvider
 import androidx.navigation.fragment.findNavController
+import com.example.wishlist.NoProductIdException
 import com.example.wishlist.data.model.Review
 import com.example.wishlist.databinding.FragmentAddReviewBinding
 import com.example.wishlist.databinding.FragmentProductDetailBinding
+import com.example.wishlist.viewmodel.ProductViewModel
+import com.example.wishlist.viewmodel.ProductViewModelFactory
```
Comment (Tonyyytran) [inline app/src/main/java/com/example/wishlist/fragments/AddReviewFragment.kt]: Are we still using this factory?

File: app/src/main/java/com/example/wishlist/fragments/ProductDetailFragment.kt [hunk 5]
```diff
@@ -23,31 +25,48 @@ class ProductDetailFragment : Fragment(), OnReviewClickedListener
...
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
+
+        getViewModel()
+        product = viewModel.getProduct(productId)
+        bindButton()
+        bindRecyclerView()
+        bindProductView()
+        bindSwipeRefresh()
+
```
Comment (Tonyyytran) [inline app/src/main/java/com/example/wishlist/fragments/ProductDetailFragment.kt]: can you remove extra line?

File: app/src/main/java/com/example/wishlist/fragments/ShoppingCartFragment.kt [hunk 6]
```diff
@@ -1,17 +1,21 @@
...
+package com.example.wishlist.fragments
 
 import android.os.Bundle
 import androidx.fragment.app.Fragment
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
 import androidx.lifecycle.ViewModelProvider
+import com.example.wishlist.viewmodel.ProductViewModel
+import com.example.wishlist.viewmodel.ProductViewModelFactory
```
Comment (Tonyyytran) [inline app/src/main/java/com/example/wishlist/fragments/ShoppingCartFragment.kt]: Remove imports if unused

File: app/src/main/java/com/example/wishlist/viewmodel/ProductViewModel.kt [hunk 7]
```diff
@@ -55,4 +62,18 @@ class ProductViewModel(application: Application): AndroidViewModel(application)
...
+
+    companion object {
+        private var instance: ProductViewModel? = null
+        fun getInstance(application: Application): ProductViewModel {
+            if (instance == null) {
+                instance = ProductViewModel(application)
+            }
+            return instance!!
+        }
+}}
```
Comment (Tonyyytran) [inline app/src/main/java/com/example/wishlist/viewmodel/ProductViewModel.kt]: reformat this curly hair

File: app/src/main/res/layout/fragment_shopping_cart.xml [hunk 8]
```diff
@@ -4,21 +4,21 @@
...
     android:layout_height="match_parent"
     xmlns:app="http://schemas.android.com/apk/res-auto"
-    tools:context=".ShoppingCartFragment"
+    tools:context=".fragments.ShoppingCartFragment"
     android:orientation="vertical">
 
     <TextView
         android:layout_width="match_parent"
-        android:layout_height="wrap_content"
+        android:layout_height="48dp"
```
Comment (Tonyyytran) [inline app/src/main/res/layout/fragment_shopping_cart.xml]: is this intentional?

File: app/src/main/java/com/example/wishlist/adapters/ShoppingCartAdapter.kt [hunk 9]
```diff
@@ -25,10 +28,15 @@ class ShoppingCartAdapter(var cartItems: List<Product> = emptyList()): RecyclerV
...
-        return cartItems.size
-    }
+class CartItemDiffCallback(): DiffUtil.ItemCallback<Product>() {
 
+    override fun areItemsTheSame(oldItem: Product, newItem: Product): Boolean {
+        return oldItem.id == newItem.id
+    }
 
+    override fun areContentsTheSame(oldItem: Product, newItem: Product): Boolean {
+        return oldItem == newItem
```
Comment (Tonyyytran) [inline app/src/main/java/com/example/wishlist/adapters/ShoppingCartAdapter.kt:40]: Does this check all attributes of the product?

**Review by Tonyyytran — COMMENTED** [review]
you did wonderful. make the updates and we are all good

File: app/src/main/java/com/example/wishlist/adapters/ProductAdapter.kt (hunk 1 above)
Comment (tritran206) [reply app/src/main/java/com/example/wishlist/adapters/ProductAdapter.kt:48]: Tested evaluation and it does check all attributes.

File: app/src/main/java/com/example/wishlist/adapters/ProductAdapter.kt (hunk 2 above)
Comment (tritran206) [reply app/src/main/java/com/example/wishlist/adapters/ProductAdapter.kt:2]: Package name has been updated.

File: app/src/main/java/com/example/wishlist/adapters/ReviewAdapter.kt (hunk 3 above)
Comment (tritran206) [reply app/src/main/java/com/example/wishlist/adapters/ReviewAdapter.kt:2]: Package name updated.

File: app/src/main/java/com/example/wishlist/fragments/AddReviewFragment.kt (hunk 4 above)
Comment (tritran206) [reply app/src/main/java/com/example/wishlist/fragments/AddReviewFragment.kt]: Imports removed

File: app/src/main/java/com/example/wishlist/fragments/ProductDetailFragment.kt (hunk 5 above)
Comment (tritran206) [reply app/src/main/java/com/example/wishlist/fragments/ProductDetailFragment.kt]: Line removed.

File: app/src/main/java/com/example/wishlist/fragments/ShoppingCartFragment.kt (hunk 6 above)
Comment (tritran206) [reply app/src/main/java/com/example/wishlist/fragments/ShoppingCartFragment.kt]: Imports removed

File: app/src/main/java/com/example/wishlist/viewmodel/ProductViewModel.kt (hunk 7 above)
Comment (tritran206) [reply app/src/main/java/com/example/wishlist/viewmodel/ProductViewModel.kt]: Code has been reformatted

File: app/src/main/res/layout/fragment_shopping_cart.xml (hunk 8 above)
Comment (tritran206) [reply app/src/main/res/layout/fragment_shopping_cart.xml]: updated

File: app/src/main/java/com/example/wishlist/adapters/ShoppingCartAdapter.kt (hunk 9 above)
Comment (tritran206) [reply app/src/main/java/com/example/wishlist/adapters/ShoppingCartAdapter.kt:40]: Tested the evaluation above and it does check all the attributes

**Pull request merged**